File: src/types/swarm.ts
export interface SwarmAgentConfig {
  agentId: string
  role: 'coordinator' | 'researcher' | 'executor' | 'critic' | 'synthesizer'
  model: string
  instructions: string
  weight?: number
}

export interface SwarmConfig {
  id: string
  name: string
  description: string
  goal: string
  agents: SwarmAgentConfig[]
  consensusMode: 'majority' | 'unanimous' | 'weighted'
  maxRounds: number
  timeoutMinutes: number
  createdAt: string
}

export interface SwarmMessage {
  id: string
  round: number
  agentId: string
  agentName: string
  agentRole: 'coordinator' | 'researcher' | 'executor' | 'critic' | 'synthesizer'
  content: string
  timestamp: string
  messageType: 'proposal' | 'feedback' | 'vote' | 'consensus' | 'system'
}

export interface SwarmDecision {
  round: number
  proposal: string
  votes: {
    agentId: string
    vote: 'approve' | 'reject' | 'abstain'
    reasoning: string
    confidence: number
  }[]
  result: 'approved' | 'rejected' | 'deadlocked'
  consensusReached: boolean
}

export interface SwarmExecution {
  id: string
  configId: string
  status: 'running' | 'paused' | 'completed' | 'failed' | 'stopped'
  currentRound: number
  maxRounds: number
  startTime: string
  endTime?: string
  messages: SwarmMessage[]
  decisions: SwarmDecision[]
  finalOutput?: string
  cost: number
}

File: backend/app/schemas/swarm.py
from pydantic import BaseModel, ConfigDict


class SwarmConfigResponse(BaseModel):
    id: str
    name: str
    description: str = ""
    goal: str
    agents: list
    consensus_mode: str = "majority"
    max_rounds: int = 5
    timeout_minutes: int = 60
    created_at: str

    model_config = ConfigDict(from_attributes=True)


class SwarmExecutionResponse(BaseModel):
    id: str
    config_id: str
    status: str
    current_round: int
    max_rounds: int
    start_time: str
    end_time: str | None = None
    messages: list = []
    decisions: list = []
    final_output: str | None = None
    cost: float = 0.0

    model_config = ConfigDict(from_attributes=True)
